distributed/deploy/spec.py

```diff
@@ -0,0 +1,296 @@
+    form.
+
+    >>> worker = {'cls': Worker, 'options': {}}
+    >>> cluster = SpecCluster(scheduler=scheduler, worker=worker)
+    >>> cluster.worker_spec
+    {}
+
+    >>> cluster.scale(3)
+    >>> cluster.worker_spec
+    {
+        0: {'cls': Worker, 'options': {}},
+        1: {'cls': Worker, 'options': {}},
+        2: {'cls': Worker, 'options': {}},
+    }
+
+    Note that above we are using the standard ``Worker`` and ``Nanny`` classes,
+    however in practice other classes could be used that handle resource
+    management like ``KubernetesPod`` or ``SLURMJob``.  The spec does not need
+    to conform to the expectations of the standard Dask Worker class.  It just
+    needs to be called with the provided options, support ``__await__`` and
+    ``close`` methods and the ``worker_address`` property..
+
+    Also note that uniformity of the specification is not required.  Other API
+    could be added externally (in subclasses) that adds workers of different
+    specifications into the same dictionary.
+    """
+
+    def __init__(
+        self,
+        workers=None,
+        scheduler=None,
+        worker=None,
+        asynchronous=False,
+        loop=None,
+        silence_logs=False,
+    ):
+        self._created = weakref.WeakSet()
+        if scheduler is None:
+            try:
+                from distributed.bokeh.scheduler import BokehScheduler
+            except ImportError:
+                services = {}
+            else:
+                services = {("bokeh", 8787): BokehScheduler}
+            scheduler = {"cls": Scheduler, "options": {"services": services}}
+
+        self.scheduler_spec = scheduler
+        self.worker_spec = workers or {}
+        self.new_spec = worker
+        self.workers = {}
+        self._i = 0
+        self._asynchronous = asynchronous
+
+        if silence_logs:
+            self._old_logging_level = silence_logging(level=silence_logs)
+
+        self._loop_runner = LoopRunner(loop=loop, asynchronous=asynchronous)
+        self.loop = self._loop_runner.loop
+
+        self.scheduler = self.scheduler_spec["cls"](
+            loop=self.loop, **self.scheduler_spec["options"]
+        )
+        self.status = "created"
+
+        if not self.asynchronous:
+            self._loop_runner.start()
+            self.sync(self._start)
+
+        self._correct_state_waiting = None
+
+    async def _start(self):
+        while self.status == "starting":
+            await asyncio.sleep(0.01)
+        if self.status == "running":
+            return
+        if self.status == "closed":
+            raise ValueError("Cluster is closed")
+
+        self._lock = asyncio.Lock()
+        self.status = "starting"
+        self.scheduler = await self.scheduler
+        self.status = "running"
+
+    def _correct_state(self):
+        if self._correct_state_waiting:
+            # If people call this frequently, we only want to run it once
```

Comment:
I think this drops scale requests while a current scale request is processing:

- Call scale
- Spec updated
- correct state task start, task is stored as `_correct_state_waiting`
- scale returns
- Call scale again
- Spec updated
- since previous call is still in progress, state is not corrected, no new workers are started/stopped. Spec and tasks are now out of sync. Also, since there are multiple `await` calls in `_correct_state_internal`, the `worker_spec` can be different at different points in that function, leading to potential bugs.

One naive solution would be to have a background task that loops forever, waiting on an event:


```python
while self.running:
    await self._spec_updated.wait()
    # update workers to match spec
    # After updating, only clear the event if things are up to date
    # If things aren't up to date, then we loop again
    if self.spec_matches_current_state():
        self._spec_updated.clear()
```

Then `_correct_state` would look like:

```python
def _correct_state(self):
    # set the event, it's only ever cleared in the loop
    # We force synchronization here to prevent scheduling tons
    # of tasks all setting the event, this blocks until it's set.
    return self.sync(self._mark_state_updated)

async def _mark_state_updated(self):
    self._state_updated.set()
```

There are likely other ways to handle this. In dask-gateway I have a task per worker/scheduler. As the spec updates, unfinished tasks are cancelled or new ones are fired. If a previous `scale` call is still in progress for a cluster, `scale` will block until that call has finished. Note that this only blocks while we update our internal task state (cancelling/firing new tasks), not until those tasks have completed.

Reply:
> since previous call is still in progress, state is not corrected, no new workers are started/stopped. Spec and tasks are now out of sync. Also, since there are multiple await calls in _correct_state_internal, the worker_spec can be different at different points in that function, leading to potential bugs.

So, the `_correct_state_waiting` attribute isn't the currently running task, it's the currently enqueued one.  Once `_correct_state` starts running it immediately clears this attribute.  After someone calls `scale` there is a clean, not-yet-run `_correct_state_waiting` future that will run soon.

Reply:
Since `_correct_state_waiting` waits on the created workers, this does mean that there's no way to cancel pending workers. This is fine for `LocalCluster`, but would be problematic if used as a base class for other cluster managers. The following would request and start 100 workers before scaling back down afaict:

```python
cluster.scale(100)
cluster.scale(2)
```

Reply:
I think that this depends on what you mean by "waits on".  

One approach is that for a cluster manager to reach a correct state it only has to successfully submit a request to the resource manager have received an acknowledgement that the resource manager is handling it. We're not guaranteeing full deployment, merely that we've done our part of the job.  I would expect this to almost always be fairly fast.

Separately, there is now a `Client.wait_for_workers(n=10)` method that might be used for full client <-> scheduler checks.